app/controllers/agents_controller.rb

```diff
@@ -220,9 +220,9 @@ def redirect_back(message, options = {})
   end
 
   def build_agent
-    @agent = Agent.build_for_type(params[:agent].delete(:type),
+    @agent = Agent.build_for_type(agent_params.delete(:type),
```

Comment:
You have `agent_params` memoized, so this should work the same way it did before, but I think it was written in a confusing way.  Could we do:

```ruby
params_without_type = agent_params.except(:type)
@agent = Agent.build_for_type(params_without_type,
                                                    current_user,
                                                    params_without_type)
```